File: files/300-399/338/CH-338.py
import pandas as pd
import re

path = "300-399\\338\\CH-338 Column Splitting.xlsx"
input = pd.read_excel(path, usecols="B", nrows=5, skiprows=2)
test = pd.read_excel(path, usecols="F:K", nrows=5, skiprows=2)

colname = input.columns[0]

def add_splitter_on_char_change(x):
    pattern = (
        r"(?<=[A-Za-z])(?=[^A-Za-z])|(?<=[^A-Za-z])(?=[A-Za-z])|"
        r"(?<=[0-9])(?=[^0-9])|(?<=[^0-9])(?=[0-9])|"
        r"(?<=[\W_])(?=[^\W_])|(?<=[^\W_])(?=[\W_])"
    )
    return re.sub(pattern, "|", str(x))

split_cols = input[colname].apply(add_splitter_on_char_change).str.split("|")
maxlen = split_cols.map(len).max()
colnames = [f"ID_{i+1}" for i in range(maxlen)]
result = pd.DataFrame(split_cols.tolist(), columns=colnames)

print(result.equals(test)) # one field incorrect in answers provided

File: files/300-399/338/CH-338.R
library(tidyverse)
library(readxl)

path <- "300-399/338/CH-338 Column Splitting.xlsx"
input <- read_excel(path, range = "B3:B8")
test <- read_excel(path, range = "F3:K8")

add_splitter_on_char_change <- function(x) {
  stringr::str_replace_all(
    x,
    "(?<=[A-Za-z])(?=[^A-Za-z])|(?<=[^A-Za-z])(?=[A-Za-z])|
     (?<=[0-9])(?=[^0-9])|(?<=[^0-9])(?=[0-9])|
     (?<=[[:punct:]])(?=[^[:punct:]])|(?<=[^[:punct:]])(?=[[:punct:]])",
    "|"
  )
}

result = input %>%
  mutate(ID = map(ID, ~ add_splitter_on_char_change(.))) %>%
  separate_wider_delim(
    ID,
    delim = "|",
    names_sep = " ",
    too_few = "align_start"
  )

all.equal(result, test)
